android/src/ti/identity/FingerPrintHelper.java

```diff
@@ -78,7 +79,14 @@ public FingerPrintHelper()
 	protected boolean isDeviceSupported()
 	{
 		if (Build.VERSION.SDK_INT >= 23 && mBiometricManager != null) {
-			return mBiometricManager.canAuthenticate() == BiometricManager.BIOMETRIC_SUCCESS;
+			if (mBiometricManager.canAuthenticate() == BiometricManager.BIOMETRIC_SUCCESS) {
+				mDeviceCredential = false;
+				return true;
+			} else if (mBiometricManager.canAuthenticate() == BiometricManager.BIOMETRIC_ERROR_NO_HARDWARE
+					   || mBiometricManager.canAuthenticate() == BiometricManager.BIOMETRIC_ERROR_NONE_ENROLLED) {
+				mDeviceCredential = true;
+				return true;
+			}
```

Comment:
This isn't quite right. We actually need to call `mKeyguardManager.isDeviceSecure()` to check if device the device has a passcode set up. Also, I'd prefer that we not store the state to member variable `mDeviceCredential` in case the state changes unexpectedly.

In order to check this properly, we need to change `FingerPrintHelper` to fetch the policy setting the module is currently configured with. We can do this by having the `TitaniumIdentityModule` class pass an instance of itself to `FingerPrintHelper` via its constructor like this.
```java
private TitaniumIdentityModule mModule;

public FingerPrintHelper(TitaniumIdentityModule module)
{
	if (module == null) {
		throw new NullPointerException();
	}

	mModule = module;

	// The rest of the constructor code goes here...
}
```

You can then handle `isDeviceSupported()` like the below and add additional methods to determine if device biometrics or credentials is available. The last 2 methods can be used by the `startListening()` method to determine how to authenticate the end-user.
```java
protected boolean isDeviceSupported()
{
	return canUseDeviceBiometrics() || canUseDeviceCredentials();
}

private boolean canUseDeviceBiometrics()
{
	if ((Build.VERSION.SDK_INT >= 23) && (mBiometricManager != null)) {
		if (mBiometricManager.canAuthenticate() == BiometricManager.BIOMETRIC_SUCCESS) {
			return true
		}
	}
	return false;
}

private boolean canUseDeviceCredentials()
{
	if ((Build.VERSION.SDK_INT >= 23) && (mBiometricManager != null)) {
		if (mModule.getAuthenticationPolicy() == TitaniumIdentityModule.AUTHENTICATION_POLICY_PASSCODE) {
			try {
				return mKeyguardManager.isDeviceSecure();
			} catch (Exception ex) {
			}
		}
	}
	return false;
}
```

Reply:
Done as suggested.